lib/Util/Conditions.cpp

```diff
@@ -40,51 +40,149 @@ CondExpr* CondManager::falseCond = nullptr;
 /// Constructor
 CondManager::CondManager() : sol(cxt)
 {
-    trueCond = new CondExpr(cxt.bool_val(true));
-    falseCond = new CondExpr(cxt.bool_val(false));
+    const z3::expr &trueExpr = cxt.bool_val(true);
+    trueCond = new CondExpr(trueExpr);
+    allocatedConds[trueExpr.id()] = trueCond;
+    const z3::expr &falseExpr = cxt.bool_val(false);
+    falseCond = new CondExpr(falseExpr);
+    allocatedConds[falseExpr.id()] = falseCond;
 }
 
 /// Destructor
 CondManager::~CondManager()
 {
-    delete trueCond;
-    delete falseCond;
-    for (CondExpr *cond : allocatedConds)
+    for (const auto& it : allocatedConds)
     {
-        delete cond;
+        delete it.second;
     }
 }
 
+/// Simplify
+z3::expr CondManager::simplify(const z3::expr& expr){
+    z3::goal g(cxt);
+    z3::tactic qe(cxt, "ctx-solver-simplify");
+    g.add(expr);
+    z3::apply_result r = qe(g);
+    z3::goal result_goal = r[0];
+    const z3::expr &asExpr = result_goal.as_expr();
+    z3::expr res(cxt.bool_val(false));
+    for (u32_t i = 0; i < r.size(); ++i) {
+        if (res.is_false()) {
+            res = r[i].as_expr();
+        } else {
+            res = res || r[i].as_expr();
+        }
+    }
+    return res;
+}
+
 /// And operator for two expressions
 CondExpr* CondManager::AND(const CondExpr *lhs, const CondExpr *rhs)
 {
-    CondExpr *cond = new CondExpr(lhs->getExpr() && rhs->getExpr());
-    allocatedConds.insert(cond);
-    return cond;
+    if (lhs->getExpr().is_false() || rhs->getExpr().is_false())
+        return falseCond;
+    else if (lhs->getExpr().is_true()){
+        assert(allocatedConds.count(rhs->getExpr().id()) && "rhs not in allocated conds");
+        return allocatedConds[rhs->getExpr().id()];
+    } else if (rhs->getExpr().is_true()){
+        assert(allocatedConds.count(lhs->getExpr().id()) && "lhs not in allocated conds");
+        return allocatedConds[lhs->getExpr().id()];
+    } else {
+        const z3::expr& expr = simplify(lhs->getExpr() && rhs->getExpr());
```

Comment:
implement operator& | and ~ in your CondExpr class to move the following code in.